### PR TITLE
feat(ethers): restores RelayClient's _prepareRelayHttpRequest

### PR DESCRIPTION
**Pull request by jurajpiar**
## What

- adds RelayClient
- adds EventManager
- re-implements _prepareRelayHttpRequest method
- removes @nomiclabs/hardhat-ethers as not used to my knowledge
- adds test-only lint rule override (unbound-method)
- adds more precise enveloping request typing, based on typechain-generated types

- when resolving conflicts it was necessary to clean up some previous code, which also lead to moving some code to better suited locations

- one thing to note is that the whole RelayedTransactionValidator needs to be reworked to remove relay common as dependency ([PP-618](https://rsklabs.atlassian.net/browse/PP-618))

## Refs

- [ticket](https://rsklabs.atlassian.net/browse/PP-604)

### CODE REVIEW

File: src/RelayClient.ts [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+    request: EnvelopingTypes.RelayRequestStruct['request'],
+    data: EnvelopingTypes.RelayRequestStruct['relayData'],
+    requestConfig: RequestConfig
+  ): Promise<PreparedRelayHttpRequest> => {
+    const forwarderAddress =
+      config.get<string>('forwarderAddress') ??
+      data.callForwarder ??
+      constants.AddressZero;
+
+    const provider = ethers.getDefaultProvider();
```
Comment (franciscotobar) [inline src/RelayClient.ts:77]: Where are we going to obtain or configure the RSK node address? Should we get it from the server or the configuration?

**Comment by franciscotobar** [comment]
Missing unit testing

File: src/RelayClient.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+  clientId?: string;
+  useEnveloping?: boolean;
+  forceGasPrice?: string;
+  forceGas?: string;
+  onlyPreferredRelays?: boolean;
+  ignoreTransactionReceipt?: boolean;
+  retries?: number;
+  initialBackoff?: number;
+};
+type PreparedRelayHttpRequest = {
```
Comment (franciscotobar) [inline src/RelayClient.ts:52]: based on this type, we are changing the request that would be handled by the relayServer

File: src/RelayClient.ts [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+): request is EnvelopingTypes.DeployRequestStruct => {
+  return deployOnlyRequestKeys.every((key: DeployOnlyRequestKey) => {
+    return key in request;
+  });
+};
+
+type RequestConfig = {
+  isSmartWalletDeploy?: boolean;
+  smartWalletAddress?: string;
+  clientId?: string;
```
Comment (franciscotobar) [inline src/RelayClient.ts:43]: I dont see any use for the clientId

File: src/RelayClient.ts [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+const isDeployRequest = (
+  request: EnvelopingTypes.DeployRequestStruct
+): request is EnvelopingTypes.DeployRequestStruct => {
+  return deployOnlyRequestKeys.every((key: DeployOnlyRequestKey) => {
+    return key in request;
+  });
+};
+
+type RequestConfig = {
+  isSmartWalletDeploy?: boolean;
```
Comment (franciscotobar) [inline src/RelayClient.ts:41]: do we need this if we already have a way to validate the request type?

File: src/RelayClient.ts [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  private _prepareRelayHttpRequest = async (
+    { feesReceiver }: HubInfo,
+    request: EnvelopingTypes.RelayRequestStruct['request'],
+    data: EnvelopingTypes.RelayRequestStruct['relayData'],
+    requestConfig: RequestConfig
+  ): Promise<PreparedRelayHttpRequest> => {
+    const forwarderAddress =
+      config.get<string>('forwarderAddress') ??
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/RelayClient.ts:73]: I'll adapt. thanks

File: src/RelayClient.ts (hunk 2 above)
Comment (jurajpiar) [reply src/RelayClient.ts:52]: Yea.. is it :/ was hoping it wouldn't be too bad to change...would it?
I put it in the description now.

File: src/RelayClient.ts (hunk 4 above)
Comment (jurajpiar) [reply src/RelayClient.ts:41]: just following the original implementation. Should I remove it, do you reckon?

File: src/RelayClient.ts (hunk 3 above)
Comment (jurajpiar) [reply src/RelayClient.ts:43]: just following the original implementation. Should I remove it, do you reckon?

File: src/RelayClient.ts (hunk 4 above)
Comment (franciscotobar) [reply src/RelayClient.ts:41]: Ill wait for @antomor opinion.

File: src/RelayClient.ts (hunk 3 above)
Comment (franciscotobar) [reply src/RelayClient.ts:43]: Ill wait for @antomor opinion.

File: src/RelayClient.ts [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    this._provider = getDefaultProvider();
+    this._envelopingConfig = getEnvelopingConfig();
+  }
+
+  // TODO: lift the ts exception once method is used
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  private _prepareRelayHttpRequest = async (
+    { feesReceiver }: HubInfo,
+    request: EnvelopingTypes.RelayRequestStruct['request'],
```
Comment (franciscotobar) [inline src/RelayClient.ts:76]:
```suggestion
   { request, relayData }: EnvelopingTypes.RelayRequestStruct,
```

Comment (franciscotobar) [reply src/RelayClient.ts:76]: shouldn't be better to keep the relayRequest together and just destructuring?

**Comment by franciscotobar** [comment]
Maybe is out of the scope, but it would be better to have the `_prepareDeployHttpRequest` and `_prepareRelayHttpRequest` as a single method?

File: src/RelayClient.ts [hunk 7]
```diff
@@ -0,0 +1,140 @@
+import type { RelayMetadata } from '@rsksmart/rif-relay-common';
```
Comment (franciscotobar) [inline src/RelayClient.ts:1]: This needs to be remove

File: src/RelayClient.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/RelayClient.ts]: done

File: src/RelayClient.ts (hunk 7 above)
Comment (jurajpiar) [reply src/RelayClient.ts]: done

File: src/RelayClient.ts [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+      forwarderAddress,
+      this._provider
+    );
+    const updatedRequest: EnvelopingTypes.RelayRequestStruct['request'] = {
+      ...request,
+      relayHub:
+        request.relayHub ??
+        this._envelopingConfig.relayHubAddress ??
+        constants.AddressZero,
+      tokenAmount: request.tokenAmount ?? constants.Zero,
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply src/RelayClient.ts]: done

File: src/RelayClient.ts (hunk 6 above)
Comment (jurajpiar) [reply src/RelayClient.ts]: resolved

File: src/RelayClient.ts (hunk 2 above)
Comment (jurajpiar) [reply src/RelayClient.ts]: removed type

File: test/RelayClient.test.ts [hunk 9]
```diff
@@ -0,0 +1,316 @@
+/* eslint-disable @typescript-eslint/no-unused-vars */
```
Comment (franciscotobar) [inline test/RelayClient.test.ts:1]: we need to remove this

File: test/RelayClient.test.ts [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+import EnvelopingEventEmitter, {
+  envelopingEvents,
+} from '../src/events/EnvelopingEventEmitter';
+import RelayClient from '../src/RelayClient';
+
+use(sinonChai);
+use(chaiAsPromised);
+const sandbox = Sinon.createSandbox();
+
+type RelayClientExposed = {
```
[the remark this answers is not in the document]
Comment (jurajpiar) [reply test/RelayClient.test.ts:39]: erm, ok, will do

File: test/RelayClient.test.ts (hunk 9 above)
Comment (jurajpiar) [reply test/RelayClient.test.ts]: done

File: test/RelayClient.test.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (jurajpiar) [reply test/RelayClient.test.ts]: done

**Review by franciscotobar — APPROVED** [review]
LGTM

**Comment by franciscotobar** [comment]
Just missing the rebase :)

**Review by franciscotobar — APPROVED** [review]

**Pull request merged**